anap1003/assignment4/Assignment4.java

```diff
@@ -0,0 +1,67 @@
+package assignment4;
+
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+public class Assignment4 {
+
+    private boolean areAdjacent(int tile1, int tile2, int cols) {
+        int row1 = tile1 / cols;
+        int col1 = tile1 % cols;
+        int row2 = tile2 / cols;
+        int col2 = tile2 % cols;
+        return row1 == row2 && Math.abs(col1 - col2) == 1 || col1 == col2 && Math.abs(row1 - row2) == 1;

```

Comment:
When you have a 'clever' line of code like this. Please make sure that you add in additional brackets to make it obvious what you are doing. You are relying on && being earlier in operational order than ||; and many programmers don't know that off by heart.